File: Code/java/vector-api/slice.java
import java.util.Arrays;
import java.util.stream.IntStream;
import jdk.incubator.vector.*;

public class slice {
  public static final VectorSpecies<Integer> SPECIES = IntVector.SPECIES_PREFERRED;

  public static void micro(int origin, int [] src1, int [] src2, int [] res) {
     for (int i = 0; i < res.length; i += SPECIES.length()) {
         IntVector vec1 = IntVector.fromArray(SPECIES, src1, i);
         IntVector vec2 = IntVector.fromArray(SPECIES, src2, i);
         vec1.slice(4, vec2).intoArray(res, i);
     }
  }

  public static void main(String [] args) {
     int [] res = new int[1024];
     int origin =  Integer.parseInt(args[0]);
     var src1 = IntStream.range(0, 1024).toArray();
     var src2 = IntStream.range(0, 1024).map(i -> i * 10).toArray();

     // Warming up profile for multiple receiver types (128 and 256 bit vectors).
     for (int i = 0; i < 20000; i++) {
        micro(origin, src1, src2, res);
     }

     // Main measurement loop operates over 256 bit vector.
     long t1 = System.currentTimeMillis();
     for (int i = 0; i < 200000; i++) {
        micro(origin, src1, src2, res);
     }
     long t2 = System.currentTimeMillis();
     System.out.println("[time] " + (t2-t1) + " ms");
     System.out.println(Arrays.toString(Arrays.copyOfRange(res, 0, 16)));
  }
}

File: Code/java/vector-api/slice.scala
import jdk.incubator.vector.IntVector
import java.util.Arrays

object slice {
  val SPECIES = IntVector.SPECIES_256

  def micro(origin: Int, src1 : Array[Int] , src2 : Array[Int] , res : Array[Int]) {
     var vec1 = IntVector.fromArray(SPECIES, src1, 0)
     var vec2 = IntVector.fromArray(SPECIES, src1, 0)
     vec1.slice(origin, vec2)
         .intoArray(res, 0)
  }

  def main(args : Array[String]) {
     var i : Int = 0
     var elem : Int = 0
     var origin =  Integer.parseInt(args(0))
     var src1 = new Array[Int](16)
     var src2 = new Array[Int](16)
     var res  = new Array[Int](16)
     Arrays.fill(src1, 10)
     Arrays.fill(src2, 20)
     for (i <- 0 to 10000) {
        micro(origin, src1, src2, res) 
     }
     for (elem <- res) {
       println(elem)
     }
  }
}
